enterprise/server/remote_execution/container/container.go

```diff
@@ -243,6 +251,16 @@ func PullImageIfNecessary(ctx context.Context, env environment.Env, ctr CommandC
 		})
 		return ctr.PullImage(ctx, creds)
 	}
+
+	// TODO(iain): the auth/existence/pull synchronization is getting unruly.
+	uncastmu, _ := pullOperations.LoadOrStore(ctr.IsolationType()+imageRef, &sync.Mutex{})
```

Comment:
The `sync.Map` usage here is a memory leak since map entries are never deleted, and the blocking on `mu.Lock` doesn't respect context cancellation / timeout.

Can we use the `singleflight` package for this instead? (example usage in `ociconv.go`)

Reply:
discussed offline, singleflight is not quite what we want because we have to also think about creds. I don't think the memory buildup will be noticeable in practice so this approach lgtm